File: 15-3sum/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a, b) => a - b)
    let res = []

    for(let i = 0; i < nums.length; i++)
    {
        if (i > 0 && nums[i] === nums[i - 1]) continue

        let l = i + 1, r = nums.length - 1
        while (l < r)
        {
            total = nums[i] + nums[l] + nums[r]

            if (total < 0) l ++
            else if (total > 0) r --
            else {
                res.push([nums[i], nums[l], nums[r]])
                l ++
                while (nums[l] === nums[l - 1] && l < r)
                    l ++
            }          
        }
    }

    return res
};

File: 15-3sum/3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums.sort()

        for i, num in enumerate(nums):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            l, r = i + 1, len(nums) - 1
            while l < r:
                total = num + nums[l] + nums[r]
                if total > 0:
                    r -= 1
                elif total < 0:
                    l += 1
                else:
                    res.append([num, nums[l], nums[r]])
                    l += 1
                    while l < r and nums[l - 1] == nums[l]:
                        l += 1
            
        return res
            
                

        